Refactor of Documentable model and add Filter view

romanowski opened the pull request:
In this PR I've refactor the documentable model. We've used a ton a extensions and reasoning about members was super hard.

In this PR all members shares common api (with, Kind, Modifiers, Signature etc.) as well as shares common way to store nested members. This PR also moves us away from dokka core model so if needed we can migrate away from dokka and keep most logic intact.

Add option to fiter members that is quite generic.

abgruszecki commented on src/main/scala/dotty/renderers/ScalaHtmlRenderer.scala:
```diff
@@ -40,10 +41,69 @@ class ScalaHtmlRenderer(ctx: DokkaContext) extends SiteRenderer(ctx) {
         }
     }
 
+    private val clazz = `class`
```
Minor nitpick: this is a bit confusing, since one most often sees `clazz` variables when dealing with JVM reflection. The `class` attribute is also available as `cls`, so let's maybe use that instead?

romanowski replied:
I do not have strong preference, we can go with `cls`

abgruszecki reviewed (approved):
If we want to merge this to avoid conflicts, then let's do it.

romanowski committed:
Refactor documentable model in dokka

In this PR I've refactor the documentable model. We've used a ton a extensions and reasoning about members was super hard.

In this PR all members shares common api (with, Kind, Modifiers, Signature etc.) as well as shares common way to store nested members. This commit also moves us away from dokka core model so if needed we can migrate away from dokka and keep most logic intact.

Add option to fiter members that is quite generic.

zgrybus committed:
Implement frontend part of FilterView

romanowski committed:
Fix introduced problems

pikinier20 commented on src/main/scala/dotty/dokka/model/api/api.scala:
```diff
@@ -0,0 +1,139 @@
+  case Type(concreate: Boolean, opaque: Boolean) extends Kind("Type") // should we handle opaque as modifier?
+  case Given(as: Option[Signature], conversion: Option[ImplicitConversion]) extends Kind("Given") with ImplicitConversionProvider
+  case Implicit(kind: Kind, conversion: Option[ImplicitConversion]) extends Kind(kind.name)  with ImplicitConversionProvider
+  case Uknown extends Kind("Unknown")
```
typo here

pikinier20 commented on src/main/scala/dotty/dokka/model/api/api.scala:
```diff
@@ -0,0 +1,139 @@
+
+  private[api] def memberExt = MemberExtension.getFrom(member)
+
+  private[api] def compisteMemberExt = CompositeMemberExtension.getFrom(member)
```
typo

pikinier20 commented on src/main/scala/dotty/dokka/model/compat.scala:
```diff
@@ -0,0 +1,50 @@
+import collection.JavaConverters._
+
+extension (m: DModule):
+  def updatePackanges(op: Seq[DPackage] => Seq[DPackage]): DModule = 
```
typo

pikinier20 commented on src/main/scala/dotty/dokka/transformers/migration.scala:
```diff
@@ -0,0 +1,39 @@
+// import dotty.dokka.put
+
+
+// object MigrationUtils:
```
What is purpose of this file?

romanowski replied:
left over from migration process...

pikinier20 reviewed (changes requested):
I've found few bugs/cosmetics that I think should be repaired:

- Implicit comment looks strange: ![image](https://user-images.githubusercontent.com/48246851/96236439-562c7100-0f9c-11eb-9630-2fc3e35b7c30.png) <https://scala3doc.s3.eu-central-1.amazonaws.com/pr-198/self/scala3doc/tests/implicitConversions/-c/index.html>

- Type hierarchy is wrong and known subtypes tab is missing - Look at [example](https://scala3doc.s3.eu-central-1.amazonaws.com/pr-198/self/scala3doc/example/level2/-documentation/index.html) , it is completely different from [this](https://scala3doc.s3.eu-central-1.amazonaws.com/pr-master/self/scala3doc/example/level2/-documentation/index.html)

- Annotations are looking strange and they are printed only in member list, not in main signature: look at AnnotatedClass [here](https://scala3doc.s3.eu-central-1.amazonaws.com/pr-198/self/scala3doc/tests/annotations/index.html) and [here](https://scala3doc.s3.eu-central-1.amazonaws.com/pr-198/self/scala3doc/tests/annotations/-annotated-class/index.html)

- Enums are not shown in member list [here](https://scala3doc.s3.eu-central-1.amazonaws.com/pr-198/self/scala3doc/tests/enumSignatures/index.html)

- I don't know if we should list top-level methods from package in the LHS navi [here](https://scala3doc.s3.eu-central-1.amazonaws.com/pr-198/self/scala3doc/tests/mergedPackage/index.html)

- Filter window does not show in package pages

- Long modifier parts don't look good [here](https://scala3doc.s3.eu-central-1.amazonaws.com/pr-198/self/scala3doc/tests/classModifiers/index.html)

romanowski committed:
Address review feedback and fix bugs
